File: crowdin/src/main/java/com/crowdin/platform/CrowdinContextWrapper.java
package com.crowdin.platform;

import android.content.Context;
import android.content.ContextWrapper;
import android.view.LayoutInflater;

import com.crowdin.platform.repository.StringRepository;
import com.crowdin.platform.transformers.ViewTransformerManager;

/**
 * Main Crowdin context wrapper which wraps the context for providing another layout inflater & resources.
 */
class CrowdinContextWrapper extends ContextWrapper {

    private CrowdinLayoutInflater layoutInflater;
    private ViewTransformerManager viewTransformerManager;

    public static CrowdinContextWrapper wrap(Context context,
                                             StringRepository stringRepository,
                                             ViewTransformerManager viewTransformerManager) {
        return new CrowdinContextWrapper(context, stringRepository, viewTransformerManager);
    }

    private CrowdinContextWrapper(Context base,
                                  StringRepository stringRepository,
                                  ViewTransformerManager viewTransformerManager) {
        super(new CustomResourcesContextWrapper(base, new CrowdinResources(base.getResources(), stringRepository)));
        this.viewTransformerManager = viewTransformerManager;
    }

    @Override
    public Object getSystemService(String name) {
        if (LAYOUT_INFLATER_SERVICE.equals(name)) {
            if (layoutInflater == null) {
                layoutInflater = new CrowdinLayoutInflater(LayoutInflater.from(getBaseContext()),
                        this, viewTransformerManager, true);
            }
            return layoutInflater;
        }

        return super.getSystemService(name);
    }
}

File: crowdin/src/main/java/com/crowdin/platform/CrowdinContextWrapper.kt
package com.crowdin.platform

import android.content.Context
import android.content.ContextWrapper
import com.crowdin.platform.data.DataManager
import com.crowdin.platform.transformer.ViewTransformerManager

/**
 * Main Crowdin context wrapper which wraps the context for providing another layout inflater & resources.
 */
internal class CrowdinContextWrapper private constructor(
    base: Context,
    dataManager: DataManager,
    private val viewTransformerManager: ViewTransformerManager,
) : ContextWrapper(
        CustomResourcesContextWrapper(
            base,
            // Use application context resources to ensure we have access to all system resources
            // This prevents crashes when WebView or other system components try to inflate views
            // See: https://github.com/crowdin/mobile-sdk-android/issues/220
            // See: https://github.com/crowdin/mobile-sdk-android/issues/266
            CrowdinResources(
                base.applicationContext?.resources ?: base.resources,
                dataManager
            ),
        ),
    ) {
    override fun getSystemService(name: String): Any? {
        if (Context.LAYOUT_INFLATER_SERVICE == name) {
            return CrowdinLayoutInflater(
                context = baseContext,
                viewTransformerManager = viewTransformerManager,
            )
        }

        return super.getSystemService(name)
    }

    companion object {
        fun wrap(
            context: Context,
            dataManager: DataManager?,
            viewTransformerManager: ViewTransformerManager,
        ): Context =
            if (dataManager == null) {
                context
            } else {
                CrowdinContextWrapper(
                    context,
                    dataManager,
                    viewTransformerManager,
                )
            }
    }
}
